### PR TITLE
Simplify parsing and printing attributes of function and arguments 

### PR DESCRIPTION
**Pull request by mattdamon108**
Regarding issue https://github.com/rescript-lang/rescript-compiler/issues/5735

This PR improves the parsing and printing attributes of functions and arguments so that it makes the implementation simple and clear for future development extensibility.

### CODE REVIEW

**Comment by mattdamon108** [comment]
Parsing looks okay now, but the printing error is due to an infinite loop.
```rescript
let f12 = @a (@b x) => 3
let f13 = @a @b (~x) => 3
```
```
  structure_item (tests/printer/expr/asyncAwait.res[97,2562+0]..[97,2562+24])
    Pstr_value Nonrec
    [
      <def>
        pattern (tests/printer/expr/asyncAwait.res[97,2562+4]..[97,2562+7])
          Ppat_var "f12" (tests/printer/expr/asyncAwait.res[97,2562+4]..[97,2562+7])
        expression (tests/printer/expr/asyncAwait.res[97,2562+13]..[97,2562+24])
          attribute "a" (tests/printer/expr/asyncAwait.res[97,2562+10]..[97,2562+12])
            []
          Pexp_fun
          Nolabel
          None
          pattern (tests/printer/expr/asyncAwait.res[97,2562+17]..[97,2562+18])
            attribute "b" (tests/printer/expr/asyncAwait.res[97,2562+14]..[97,2562+16])
              []
            Ppat_var "x" (tests/printer/expr/asyncAwait.res[97,2562+17]..[97,2562+18])
          expression (tests/printer/expr/asyncAwait.res[97,2562+23]..[97,2562+24])
            Pexp_constant PConst_int (3,None)
    ]
  structure_item (tests/printer/expr/asyncAwait.res[98,2587+0]..[98,2587+25])
    Pstr_value Nonrec
    [
      <def>
        pattern (tests/printer/expr/asyncAwait.res[98,2587+4]..[98,2587+7])
          Ppat_var "f13" (tests/printer/expr/asyncAwait.res[98,2587+4]..[98,2587+7])
        expression (tests/printer/expr/asyncAwait.res[98,2587+16]..[98,2587+25])
          attribute "a" (tests/printer/expr/asyncAwait.res[98,2587+10]..[98,2587+12])
            []
          attribute "b" (tests/printer/expr/asyncAwait.res[98,2587+13]..[98,2587+15])
            []
          Pexp_fun
          Labelled "x"
          None
          pattern (tests/printer/expr/asyncAwait.res[98,2587+17]..[98,2587+19])
            attribute "ns.namedArgLoc" (tests/printer/expr/asyncAwait.res[98,2587+18]..[98,2587+19])
              []
            Ppat_var "x" (tests/printer/expr/asyncAwait.res[98,2587+17]..[98,2587+19])
          expression (tests/printer/expr/asyncAwait.res[98,2587+24]..[98,2587+25])
            Pexp_constant PConst_int (3,None)
    ]
```

**Comment by cristianoc** [comment]
The body returned should be "3" in both cases.

**Comment by cristianoc** [comment]
Needs to *always* make progress, and put together a group of args.

**Comment by mattdamon108** [comment]
> expression (tests/printer/expr/asyncAwait.res[97,2562+23]..[97,2562+24])
>             Pexp_constant PConst_int (3,None)

```
          expression (tests/printer/expr/asyncAwait.res[97,2562+23]..[97,2562+24])
            Pexp_constant PConst_int (3,None)
```
I think this is the body returns "3"

File: tests/printer/expr/expected/asyncAwait.res.txt [hunk 1]
```diff
@@ -116,5 +116,5 @@ let f9 = x => async (~y) => 3
 let f10 = x => async y => 3
 let f11 = (. ~x) => (. ~y) => 3
 
-let f12 = @a x => 3
-let f13 = (@a @b ~x) => 3
+let f12 = @a (@b x) => 3
+let f13 = @a @b (~x) => 3
```
Comment (mattdamon108) [inline tests/printer/expr/expected/asyncAwait.res.txt:120]: It looks printing okay now.

File: tests/printer/expr/expected/asyncAwait.res.txt [hunk 2]
```diff
@@ -106,7 +106,7 @@ let _ = await {
 
 let f1 = async (~x, ~y) => x + y
 let f2 = async (@foo ~x, @bar ~y) => x + y
-let f3 = async (@bar @foo ~x as @zz z, ~y) => x + y
+let f3 = @foo async (~x as @bar @zz z, ~y) => x + y
```
Comment (mattdamon108) [inline tests/printer/expr/expected/asyncAwait.res.txt:109]: I'm not sure about this expression `@attr1 ~x as @attr2 y`.
If we parse the attributes inside `(...)`, the attributes are going to be `ppat_attributes`, the printer emits `~x as @attr1 @attr2 y` What do you think?

File: tests/ppx/react/expected/topLevel.res.txt [hunk 3]
```diff
@@ -5,10 +5,13 @@ module V3 = {
 
   @react.component
   let make =
-    (@warning("-16") ~a, @warning("-16") ~b, _) => {
-      Js.log("This function should be named 'TopLevel.react'")
-      ReactDOMRe.createDOMElementVariadic("div", [])
-    }
+    @warning("-16")
```
Comment (mattdamon108) [inline tests/ppx/react/expected/topLevel.res.txt:8]: This warning is added to `pexp_attributes` by JSX v3.

Comment (cristianoc) [reply tests/ppx/react/expected/topLevel.res.txt:8]: Parsing and printing should not change anything for JSX right?

Comment (mattdamon108) [reply tests/ppx/react/expected/topLevel.res.txt:8]: Actually, this is printing change. The ast is not changed.

Comment (mattdamon108) [reply tests/ppx/react/expected/topLevel.res.txt:8]: I don’t think the printing changes of JSX output would be a problem unless the ast is same. What do you think?
I’ll check the js ouput for this case.

Comment (cristianoc) [reply tests/ppx/react/expected/topLevel.res.txt:8]: JSX is never printed

**Comment by mattdamon108** [comment]
I ran a test of the compiler with the syntax submodule checked out to this branch. The test result looks fine.
```
$ make test-all
```

File: src/res_core.ml [hunk 4]
```diff
@@ -1534,12 +1534,6 @@ and parseParameter p =
   then
     let startPos = p.Parser.startPos in
     let uncurried = Parser.optional p Token.Dot in
-    (* two scenarios:
```
Comment (IwanKaramazow) [inline src/res_core.ml:1537]: Can we move this comment/mental model to the appropriate place?

Comment (mattdamon108) [reply src/res_core.ml:1537]: https://github.com/rescript-lang/syntax/pull/683/commits/f2a972059e3a25002a6181daaf682724f7cee8c6 How about this?

**Comment by mattdamon108** [comment]
Rebaed to master to resolve the confict in the test txt

File: src/res_core.ml [hunk 5]
```diff
@@ -1569,25 +1569,29 @@ and parseParameter p =
             let pat =
               let pat = Ast_helper.Pat.var ~loc (Location.mkloc lblName loc) in
               let loc = mkLoc startPos p.prevEndPos in
-              Ast_helper.Pat.constraint_ ~attrs:[propLocAttr] ~loc pat typ
+              Ast_helper.Pat.constraint_ ~attrs:([propLocAttr] @ attrs) ~loc pat
```
Comment (cristianoc) [inline src/res_core.ml:1572]: Use `::` not `@`

File: tests/printer/expr/expected/asyncAwait.res.txt (hunk 2 above)
Comment (cristianoc) [reply tests/printer/expr/expected/asyncAwait.res.txt:109]: Looks fine. There's only one place for attributes on labels.

**Review by cristianoc — COMMENTED** [review]
Starting to look at this. For post v10.1 release.
Would you rebase on master?

**Comment by mattdamon108** [comment]
Rebased to master

**Comment by cristianoc** [comment]
moved to compiler repo

**Pull request closed without merging**